include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -278,6 +291,28 @@ class EnableBatchSolver
         using value_type = typename ConcreteSolver::value_type;
         using Identity = matrix::Identity<value_type>;
         using real_type = remove_complex<value_type>;
+        using batch_dense = matrix::Dense<value_type>;
+        using batch_csr = matrix::Csr<value_type>;
+        using batch_ell = matrix::Ell<value_type>;
+
+        if (params.col_scaling.get_executor() &&
+            params.row_scaling.get_executor()) {
+            GKO_ASSERT_EQ(params.col_scaling.get_size(),
+                          system_matrix->get_common_size()[0] *
+                              system_matrix->get_num_batch_items());
+            GKO_ASSERT_EQ(params.col_scaling.get_size(),
+                          params.row_scaling.get_size());
+            if (auto mat = as<batch_dense>(system_matrix)) {
+                matrix::scale_in_place(params.col_scaling, params.row_scaling,
+                                       const_cast<batch_dense*>(mat.get()));
```

Comment:
I think this is actual UB. I guess the cleanest way is to just make the constructor take a mutable pointer for the system matrix. That would also reflect the usage more honestly.